### DISCUSSION TITLE
Few random small requests

### DISCUSSION THREAD

**Discussion by 610019610** [OpenRCT2/OpenRCT2#21680] [Feature requests]
Hello!
I played the game in the 2018-2020, then I dropped it for a while and returned just recently. I never was an active part of the community because I got a language barrier (I read english text without issues, but my english turn broken when talking or writing), but I always used to lurk and get updates about the develovment status. 

This time is different however: I returned after almost 5y and I'm noticed some issues still sitting here. My expectations were different honestly. This forced me to start a talking.

FEATURES:

- **A new optional file format for saving tracks**: .td6 got a lot of limitations and everybody here knowns it and find it a lot unconfortable nowadays. I thought developing and implementing a new format would be an high prio task; also I'm sure a project about it started (I read about it years ago and I even remember a development branch was existing), but now in **2024** still there aren't any news about. Why? New track pieces (0g rolls or whatever), new rides type (single-rail, hybrid, etc...), and the widespreadt use of hacks and tricks (just listing a few examples: track merging, track branching, using extra track pieces as supports or catwalks or whatever) uses made it essential, or nearby. Also I known a new file format should simplify the development task instead of making it overcomplex, and entirely preserve retrocompatibility; for these reasons I think just reworking the current td6 instead of adding another file format (like an already proposed .td8). 
I don't see any drawback here (assuming a developer would like it, takes this task and works on that), but just a lot of QoL improving, and a great gameplay enhancement.

- **Rides stats calculation mechanism is obsolete**. I hate to try to make coasters taking ideas from real ones (an obvious example: _The Smiler_), but getting penalties just because existing underlying math is too old for modern standards. For the sake of retrocompatibility I think a simple optional cheat or an optional entry in settings, allowing for better math calculation, would satisfy everybody without issues. I known walkaround exists (for example the `rides set` console command), but that adds complexity, and borrow many additional issues and sources of discomfort.
I'm not asking for anything complex here. There are a lot of way for making it very simple.

- **Overbanking and beyond vertical drop**: I'm surprised they wasn't implemented until now, above all the former one. Back in my old playing days, they were a top entry in every player wishlist aside other elements like 0g rolls. I known now developers aren't interested on making new sprites atm (and I'm aware a lot of sprites would be needed considering transitions, veichles etc...), but I'm just wondering and still dreaming.

I will avoid to make other requests. I just chose the simplest ones.

Thank you for your feedbacks. 
Regards.

**Comment by Gymnasiast** [comment]
> This time is different however: I returned after almost 5y and I'm noticed some issues still sitting here. My expectations were different honestly. This forced me to start a talking.

You are always free to request stuff, but remember this is project provided to you for free by unpaid volunteers. 

>  for these reasons I think just reworking the current td6 instead of adding another file format (like an already proposed .td8).

TD8 _was_ essentially a reworked TD6. We already stretched TD6 beyond what it was designed for. There is no place for more types of track element in TD6, for example. Do note that we _do_ support the newer coaster types in TD6, like the Hybrid Coaster - it has room for 255 different types of ride.

**Reply by 610019610** [reply]
> You are always free to request stuff, but remember this is project provided to you for free by unpaid volunteers. 

Sorry, maybe there was a misunderstanding. my words could sounds rough, but as I said before I got a language barrier; my grammar may looks good, but I really struggle (and sometimes I fail) on managing tones and shades of meaning. Please always take it in account.

I known OpenRCT is a great project, because mantained for free by people willing to somehow modernize a 20+ years old game with a little audience. I never forget or understimate it. I just was wishing to find the project in a much advanced state after my 5y break, and I was really curious about why it didnt happens. The ideas I was asking for were already proposed since many years ago, and they were in an active development status in 2019 as far as I remember, so I really was wondering about why something met a deadlock during my long hiatus. 

It's just a talking, where I expect a reply like:
 `"We stopped to work on .td8 because..."`
 `"We implemented zero-g rolls and large corkscrews but we didn't implement overbanking due to... "`
`"We never touched the math behind stats calculation because... "`
Just for making an example. 

I'm sure there is a valid reason behind the lackness of some features. Just I didn't figure it out and I wish to known it.

Side note: due to my language barrier I use to check and edit my sent message multile times. I hope this is not annoying.

**Reply by Gymnasiast** [reply]
I see. Don’t worry.

>  "We stopped to work on .td8 because..."

We have adopted a different method. We are first updating the game’s internal structures, so that we know what will be needed for TD8/NTDF. The work has not stopped, but it is slow and not very visible.

>  "We implemented zero-g rolls and large corkscrews but we didn't implement overbanking due to... "

These improvements have been added by X7. I don’t know why he didn’t implement overbanked turns, but if I had to guess, it’s because implementing an extra banking level is more complicated than self-contained pieces like the zero G rolls and large corkscrew.

>  "We never touched the math behind stats calculation because... "

We did a few fixes, but in general most of our users prefer to stick with the current system. Changing the algorithm will also change how a track performs, what price can be asked, and how popular it will be in the park.

**Reply by 610019610** [reply]
Thanks for this reply.



> We have adopted a different method. We are first updating the game’s internal structures, so that we know what will be needed for TD8/NTDF. The work has not stopped, but it is slow and not very visible

Glad to hear that.  I never was a so good player able to efficently reinventing the wheel at every new park/scenario, so relying on a good save format and sort of blueprints would be an huge QoL improvement for an almost-casual player like me. 


> These improvements have been added by X7. I don’t know why he didn’t implement overbanked turns, but if I had to guess, it’s because implementing an extra banking level is more complicated than self-contained pieces like the zero G rolls and large corkscrew.

I imagine implementing new levels of banking or steepness, then figure out every combo, variation and transition needed would add layers and layers of complexity. I think implementing some elements just as a stand-alone pieces with built-in transitions (as X7 did about zero-g rolls, large corckscrews or whatever) instead would be easier. That mechanism would lack flexibility but at least it's something obtainable. When I could get the following stand-alone pieces I (and surely many others) would be very happy, despite the owned limitations:
- 90° degree turn Overbanked curve (left/right)
- 45° degree turn Large overbanked curve (left/right)
- Raising/Falling overbanked curve (left/right); they are in RCT3 but I'm not remembering their names
- Hammerhead turn hat
- Wave hill hat
- Stengel dive (different sizes?)
- Overbanked S-bend (left-to-right and right-to-left); I'm not able to figuring out a better name for this piece I have in mind.
- 3/4 of Barrel roll (but maybe I'm already listing too many, so I force myself to stop here....).

Anyway I'm not really pretending anything here, but just continue the talking. I known making good RCT tiles/sprites is really hard, and I think it's the real definitive issue here. Somewhere I read X7 put too many efforts for making new pieces look decently and while reading I understood (but maybe I'm wrong) he doesn't want to implement new pieces anymore.



> We did a few fixes, but in general most of our users prefer to stick with the current system. Changing the algorithm will also change how a track performs, what price can be asked, and how popular it will be in the park.

The current system still works in many case and I'm sure nobody wants to break tons of work people made in 20+ years of playing and sharing. There's just some edge cases (but important) which could take benefits from a new formula. 
I think a second formula aside the current one, activating just by a ride flag or a cheat (but I prefer the flag idea, beause acting in the single case, instead of altering the entire scenario) would accomplish the aim without borrowing additional efforts or collateral issues. However Im not so aware about how the develop works, so I admit I might understimate the nedeed effort or something else, for example if implementing a new flag would be a non-trivial task, despite the starting idea looks as very simple.

**Reply by X123M3-256** [reply]
> These improvements have been added by X7. I don’t know why he didn’t implement overbanked turns, but if I had to guess, it’s because implementing an extra banking level is more complicated than self-contained pieces like the zero G rolls and large corkscrew.

I don't actually think it would be that hard to add an extra banking level. After I added the large sloped turns I had been planning to add overbanks next, because I thought that they would be particularly useful on the hybrid and single rail coasters, which at that time were the only coasters that got the new track pieces. I can't remember if I posted on the OpenRCT2 discord about this but there was a lot of discussion on NE. It was at one point going to be my next addition.

The actual problem was that I couldn't decide what overbanks should look like. I gathered together some pictures of different overbanks, they come in so many different sizes, shapes, and angles that it's hard to narrow it down to a handful of pieces for inclusion in the game. Some are low to the ground, some really tall, some are banked far past 90 degrees, some only at 90. Some would need to start from level track, some from sloped track, some from regular banked track. At one point, I tried to recreate some real coasters that featured overbanks to see where a new overbank piece might fit in, and came to the conclusion that every overbank on the ride would have had to be a different piece to work. It isn't practical to make 50 different overbank pieces let along come up with a clean UI for all of that.

RCT3 featured an overbank piece, but I don't like how they did it. It's inflexible because you can only make a 180 degree turn with it, and it looks weird because the coasters that had that piece available don't have overbanks that look like that in real life (the Maurer spinning coasters do have elements like that, but they are usually a lot smaller than the RCT3 piece). 

Another issue is, should the piece be truly overbanked, or should it be banked at 90 degrees? I felt that the 90 bank fits better with the game and its existing UI, since it would mean bank angles go in 45 degree angles, and it would allow the overbank pieces to fit together with other 90 banked pieces that might hypothetically be added some day, such as a double inverting stall or lagoon roll. However, some overbanks are banked well past 90 and for some ride types (like the giga) anything else looks odd. Someone made some mockups in No Limits, and I thought both versions looked good, but also that you can't really have both as it would be too many pieces. I considered having it dependent on the ride type, like how the alpine coaster gets a shallower bank angle, but this brings its own problems.

After I moved on from the idea, I had more time to think about it and I eventually decided that the following set of pieces would be best for the hybrid:

* Gentle slope 45 bank to 90 bank, 45 degree turn
* Gentle slope 45 bank to 90 bank, 90 degree small turn.
* Gentle slope 45 bank to 90 bank, 90 degree medium turn.
* Steep slope to 90 bank, straight
* Steep slope to 90 bank, 45 degree turn
* Steep slope to 90 bank, 90 degree turn.

These pieces could be combined in different ways to create 90 banked turns anywhere from 45 to 180 degrees, in two different styles. The pieces starting from a gentle slope would be suited for something like the final turn on Railblazer while the steep pieces would look more like the first turn on Mako. It allows for quite a bit of flexibility relative to the number of pieces that are being added, and I think this set of pieces would be well suited to the needs of the LIM, hybrid, and B&M ride types (although the latter two ride types can also feature overbanks banked beyond 90 degree). I wouldn't put any of the above pieces in the special menu, instead, I'd add a new 90 degree bank angle to the bank selector, so you could select an overbank piece just by selecting the appropriate turn radius and bank angle.

But still, many other styles of overbank are possible. Like the zero G rolls, this would suffer from the issue that there's a big gap between the small version that starts on gentle and the large version that starts on steep, and you'd often find yourself wanting a version that starts on a (currently nonexistent) medium slope. Particularly left out is the giga coaster - which is one of the ride types that currently gets the new pieces, and often features overbanks that are banked well past 90 degree and look very different from anything that could be created with the above pieces. One option is that since none of the above pieces would need to go in the special menu, there would be room for an "overbanked turn" special piece as well, to provide additional options - but again, I've no real idea what that piece should look like exactly.

Due to the issues, I moved on from trying to implement overbanked turns, and after the giga coaster got the new pieces and with talk of the B&M getting them too, I felt that adding new inversion pieces, particularly dive loops, should be a higher priority, but in the end I never got around to that either. Even if I had stuck with it, the fact is that there were many more track pieces I wanted to add than I had time to implement and I might never have got to this.

**Reply by X123M3-256** [reply]
> I known OpenRCT is a great project, because mantained for free by people willing to somehow modernize a 20+ years old game with a little audience. I never forget or understimate it. I just was wishing to find the project in a much advanced state after my 5y break, and I was really curious about why it didnt happens.

The project *is* in a much more advanced state. 2019 was before the new save format, before the plugin system, and before any of the new ride types or new track pieces, before custom terrain and music objects, before the new colors. A lot of *major* features have been added since then.

>  The ideas I was asking for were already proposed since many years ago, and they were in an active development status in 2019 as far as I remember, so I really was wondering about why something met a deadlock during my long hiatus.

None of the ideas you mentioned were in active development in 2019. A new track design format was always planned but there would have been no point working on that before the new save format. New track pieces were not possible to add at all before the new save format was done, nobody had even tried to add one (except for the boosters which were added with a dirty hack that woudn't work for any other track piece) Redoing the stat calculation is something I personally would have liked to see but to my knowledge, nobody has ever worked on that and a lot of people would be very opposed to that if it wasn't optional.

You seem to be under the impression that if a suggested feature hasn't been added it's because the dev team rejected it. 99% of the time, people who suggest ideas aren't actually interested in implementing them. The answer to "why hasn't this feature been added" is almost always because nobody has been working on it. If there's nobody interested in actually making the feature, it won't happen no matter how many times it has been suggested, which is why even very simple and very frequently requested features like sloped brakes have not been added yet. Even if someone is actively working on a new feature, a lot of projects never get finished. I have two new ride types that I was working on and intended to submit at the time, but which were never finished and probably never will be.
